onnxruntime/core/providers/snpe/snpe_lib.cc

```diff
@@ -0,0 +1,467 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/snpe/snpe_lib.h"
+#include <iostream>
+#include "core/common/common.h"
+#include "core/common/logging/logging.h"
+#include "DlContainer/IDlContainer.hpp"
+#include "DlSystem/ITensorFactory.hpp"
+#include "DlSystem/IUserBufferFactory.hpp"
+#include "SNPE/SNPEBuilder.hpp"
+
+namespace onnxruntime {
+namespace contrib {
+namespace snpe {
+
+size_t CalcSizeFromDims(const zdl::DlSystem::Dimension* dims, size_t rank, size_t elementSize) {
+  if (rank == 0) return 0;
+  size_t size = elementSize;
+  while (rank--) {
+    size *= *dims;
+    dims++;
+  }
+  return size;
+}
+
+bool SnpeLib::SetupUserBufferAttribute(const std::string& name) {
+  auto buffer_attributes = snpe_->getInputOutputBufferAttributes(name.c_str());
+  if (!buffer_attributes) {
+    LOGS_DEFAULT(ERROR) << "Error obtaining attributes for input/output tensor" << name;
+    return false;
+  }
+  zdl::DlSystem::UserBufferEncoding* user_buffer_encoding = nullptr;
+  size_t buffer_element_size = 0;
+  if (BufferType::TF8 == buffer_type_ || BufferType::TF16 == buffer_type_) {
+    int bit_width = buffer_type_ == BufferType::TF16 ? 16 : 8;
+    buffer_element_size = bit_width / 8;
+    auto encoding = (*buffer_attributes)->getEncoding();
+    if (!encoding) {
+      LOGS_DEFAULT(ERROR) << "Failed to get buffer encoding for: " << name;
+      return false;
+    }
+    user_buffer_encoding_.push_back(std::make_unique<zdl::DlSystem::UserBufferEncodingTfN>(*encoding));
+    user_buffer_encoding = user_buffer_encoding_.back().get();
+  } else if (BufferType::FLOAT == buffer_type_) {
+    buffer_element_size = sizeof(float);
+    user_buffer_encoding_.push_back(std::make_unique<zdl::DlSystem::UserBufferEncodingFloat>());
+    user_buffer_encoding = user_buffer_encoding_.back().get();
+  } else if (BufferType::UINT8 == buffer_type_) {
+    buffer_element_size = 1;
+    user_buffer_encoding_.push_back(std::make_unique<zdl::DlSystem::UserBufferEncodingUnsigned8Bit>());
+    user_buffer_encoding = user_buffer_encoding_.back().get();
+  }
+
+  zdl::DlSystem::Optional<zdl::DlSystem::TensorShape> tensor_shape = (*buffer_attributes)->getDims();
+  auto shape_rank = (*tensor_shape).rank();
+  std::vector<size_t> strides(shape_rank);
+  strides[strides.size() - 1] = buffer_element_size;
+  size_t stride = strides[strides.size() - 1];
+  for (size_t j = shape_rank - 1; j > 0; j--) {
+    stride *= (*tensor_shape)[j];
+    strides[j - 1] = stride;
+  }
+  size_t bufSize = CalcSizeFromDims((*tensor_shape).getDimensions(), shape_rank, buffer_element_size);
+
+  user_buffer_attr_table_.emplace(name, UserBufferAttribute(bufSize, strides, user_buffer_encoding));
+
+  return true;
+}
+
+bool SnpeLib::SetupUserBufferAttributes(const std::vector<std::string>& tensor_names) {
+  for (size_t i = 0; i < tensor_names.size(); ++i) {
+    bool rt = SetupUserBufferAttribute(tensor_names.at(i));
```

Comment:
general comment: vector::at() is checked, vector::operator[] is not. when the index is known to be in range (e.g., i < tensor_names.size()), consider just using the latter (e.g., tensor_names[i])